### PR TITLE
[MM-24495] Allow Previewing Channels

### PR DESCRIPTION
**Pull request by EricssonLiu**
#### Summary
WIP - Adds api endpoints for allowing channel previewing

- [x] add previewer
- [x] remove previewer
- [x] convert previewer
- [ ] enforce previewer attributes (hide from members list, etc..)

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-24495

#### Release Note
```release-note
Adds new endpoints for adding, removing, and converting previewers to public channels
```

### CODE REVIEW

File: api4/api.go [hunk 1]
```diff
@@ -47,6 +47,8 @@ type Routes struct {
 	ChannelsForTeam          *mux.Router // 'api/v4/teams/{team_id:[A-Za-z0-9]+}/channels'
 	ChannelMembers           *mux.Router // 'api/v4/channels/{channel_id:[A-Za-z0-9]+}/members'
 	ChannelMember            *mux.Router // 'api/v4/channels/{channel_id:[A-Za-z0-9]+}/members/{user_id:[A-Za-z0-9]+}'
+	ChannelPreviewers		 *mux.Router // 'api/v4/channels/{channel_id:[A-Za-z0-9]+}/previewers'
+	ChannelPreviewer 		 *mux.Router // 'api/v4/channels/{channel_id:[A-Za-z0-9]+}/previewer/{user_id:[A-Za-z0-9]+}'
```
Comment (ashishbhate) [inline api4/api.go:51]: The api path should be plural, in keeping with convention 

```suggestion
	ChannelPreviewer 		 *mux.Router // 'api/v4/channels/{channel_id:[A-Za-z0-9]+}/previewers/{user_id:[A-Za-z0-9]+}'
```

File: app/channel_test.go [hunk 2]
```diff
@@ -1554,6 +1554,30 @@ func TestAddUserToChannel(t *testing.T) {
 	require.Nil(t, err)
 }
 
+func TestAddChannelPreviewer(t *testing.T) {
+	th := Setup(t).InitBasic()
+	defer th.TearDown()
+
+	user := model.User{Email: strings.ToLower(model.NewId()) + "success+test@example.com", Nickname: "Darth Vader", Username: "vader" + model.NewId(), Password: "passwd1", AuthService: ""}
+	ruser, _ := th.App.CreateUser(th.Context, &user)
```
Comment (ashishbhate) [inline app/channel_test.go:1562]:
```suggestion
`ruser := th.CreateUser()`
```

The testing code has helpers to make things easier, it also does the error checks.

**Review by ashishbhate — COMMENTED** [review]
@EricssonLiu Looks good. I've done a first pass review since this is still a WIP. 
We need to add more tests to `TestAddChannelPreviewer` - e.g. for when a member already exists, and maybe a couple tests to check that the code returns errors correctly. 

I also suggest adding someone from the Suite Users team to review this code.

File: api4/channel.go [hunk 3]
```diff
@@ -1802,6 +1805,86 @@ func removeChannelMember(c *Context, w http.ResponseWriter, r *http.Request) {
...
 }
 
+func addChannelPreviewer(c *Context, w http.ResponseWriter, r *http.Request) {
+	c.RequireChannelId()
+	if c.Err != nil {
+		return
+	}
+
+	props := model.StringInterfaceFromJSON(r.Body)
+	userId, ok := props["user_id"].(string)
```
Comment (ashishbhate) [inline api4/channel.go:1815]: model/utils has a `MapFromJSON` method which does the same thing without needing a type assertion every time.

**Comment by michelengelen** [comment]
removing myself since I am not the ideal reviewer for this ... @iomodo can you take this?

**Comment by iomodo** [comment]
Removing myself from review since I'm OOO next couple of weeks. @EricssonLiu please request second review from someone else, thanks.

File: api4/channel.go [hunk 4]
```diff
@@ -1802,6 +1805,86 @@ func removeChannelMember(c *Context, w http.ResponseWriter, r *http.Request) {
...
+	if !ok || !model.IsValidId(userId) {
+		c.SetInvalidParam("user_id")
+		return
+	}
```
[the remark this answers is not in the document]
Comment (EricssonLiu) [reply api4/channel.go:1819]: Good question. What do we prefer here, should both parameters be passed through the body?

File: api4/channel.go [hunk 5]
```diff
@@ -1802,6 +1805,86 @@ func removeChannelMember(c *Context, w http.ResponseWriter, r *http.Request) {
...
+			isNewMembership = true
+		} else {
+			c.Err = model.NewAppError("addChannelPreviewer", "api.channel.get_channel_member.error", nil, "", http.StatusBadRequest)
+			return
+		}
+	}
+	if !isNewMembership {
+		c.Err = model.NewAppError("addChannelPreviewer", "api.channel.previewing_joined_channel.error", nil, "", http.StatusBadRequest)
+		return
+	}
```
[the remark this answers is not in the document]
Comment (EricssonLiu) [reply api4/channel.go:1855]: If the user is already a member, the preview channel options should not be available in the UI when browsing channels so I think its ok to error out in this scenario as it shouldn't happen.

**Comment by Willyfrog** [comment]
Hi @EricssonLiu how is it going? do you need any help with this?

**Comment by EricssonLiu** [comment]
> Hi @EricssonLiu how is it going? do you need any help with this?

Hey, @Willyfrog been on the move for a bit and just settling in now - should be able to find a chance to work on this these next couple weeks.

**Pull request closed without merging**